scripts/lib/CIME/build.py

```diff
@@ -562,10 +366,210 @@ def clean(case, cleanlist=None, clean_all=False):
+    # ANY of those are true.
+
+    use_trilinos = False if cism_use_trilinos is None else cism_use_trilinos
+    case.set_value("USE_TRILINOS", use_trilinos)
+    os.environ["USE_TRILINOS"] = stringify_bool(use_trilinos)
+
+    # Set the overall USE_ALBANY variable to TRUE if any of the
+    # *_USE_ALBANY variables are TRUE.
+    # For now, there is just the one MPASLI_USE_ALBANY variable, but in
+    # the future there may be others -- so USE_ALBANY will be true if
+    # ANY of those are true.
+
+    use_albany = stringify_bool(mpasli_use_albany)
+    case.set_value("USE_ALBANY", use_albany)
+    os.environ["USE_ALBANY"] = use_albany
+
+    # Load modules
+    case.load_env()
+
+    sharedpath = _build_checks(case, build_threaded, comp_interface,
+                               use_esmf_lib, debug, compiler, mpilib,
+                               complist, ninst_build, smp_value, model_only)
+
+    t2 = time.time()
+    logs = []
+
+    if not model_only:
+        logs = _build_libraries(case, exeroot, sharedpath, caseroot,
+                               cimeroot, libroot, lid, compiler)
+
+    if not sharedlib_only:
+        logs.extend(_build_model(build_threaded, exeroot, clm_config_opts, incroot, complist,
+                                lid, caseroot, cimeroot, compiler))
+
+    if not sharedlib_only:
+        # in case component build scripts updated the xml files, update the case object
+        case.read_xml()
+        post_build(case, logs)
+
+    t3 = time.time()
+
+    logger.info("Time spent not building: %f sec" % (t2 - t1))
+    logger.info("Time spent building: %f sec" % (t3 - t2))
+
+    return True
+
+###############################################################################
+def post_build(case, logs):
+###############################################################################
+
+    logdir = case.get_value("LOGDIR")
+
+    #zip build logs to CASEROOT/logs
+    if logdir:
+        bldlogdir = os.path.join(logdir, "bld")
+        if not os.path.exists(bldlogdir):
+            os.makedirs(bldlogdir)
+
+    for log in logs:
+        logger.debug("Copying build log %s to %s"%(log,bldlogdir))
+        with open(log, 'rb') as f_in:
+            with gzip.open("%s.gz"%log, 'wb') as f_out:
+                shutil.copyfileobj(f_in, f_out)
+        if "sharedlibroot" not in log:
+            shutil.copy("%s.gz"%log,os.path.join(bldlogdir,"%s.gz"%os.path.basename(log)))
+
+    # Set XML to indicate build complete
+    case.set_value("BUILD_COMPLETE", True)
+    case.set_value("BUILD_STATUS", 0)
+    if "SMP_VALUE" in os.environ:
+        case.set_value("SMP_BUILD", os.environ["SMP_VALUE"])
+    case.flush()
+
+    lock_file("env_build.xml")
+
+    # must ensure there's an lid
+    lid = os.environ["LID"] if "LID" in os.environ else run_cmd_no_fail("date +%y%m%d-%H%M%S")
+    save_build_provenance(case, lid=lid)
+
+###############################################################################
```

Comment:
This approach encapsulates an approach that ensure the CaseStatus gets updated even in the presence of exceptions.